tests/31sync/09archived.pl

```diff
@@ -0,0 +1,193 @@
+         ( $user, $filter_id ) = @_;
+
+         Future->needs_all(
+            matrix_create_room( $user )->on_done( sub { ( $room_id_1 ) = @_; } ),
+            matrix_create_room( $user )->on_done( sub { ( $room_id_2 ) = @_; } ),
+         );
+      })->then( sub {
+         matrix_sync( $user, filter => $filter_id );
+      })->then( sub {
+         my ($body ) = @_;
+
+         $next = $body->{next_batch};
+
+         matrix_leave_room( $user, $room_id_1 );
+      })->then( sub {
+         Future->needs_all( map {
+            matrix_send_room_message( $user, $room_id_2,
+               content => { "filler" => $_ },
+               type    => "a.made.up.filler.type",
+            )
+         } 0 .. 20 );
+      })->then( sub {
+         matrix_sync( $user, filter => $filter_id, since => $next );
+      })->then( sub {
+         my ( $body ) = @_;
+
+         my $room = $body->{rooms}{archived}{$room_id_1};
+         require_json_keys( $room, qw( event_map timeline state ));
+
+         Future->done(1);
+      });
+   };
+
+
+test "Archived rooms only contain history from before the user left",
+   requires => [qw( first_api_client can_sync )],
+
+   check => sub {
+      my ( $http ) = @_;
+
+      my ( $user_a, $filter_id_a, $user_b, $filter_id_b, $room_id, $next_b );
+
+      my $filter = {
+         room => {
+            timeline => { types => [ "m.room.message" ] },
+            state => { types => [ "a.madeup.test.state" ] },
+         },
+      };
+
+      Future->needs_all(
+         matrix_register_user_with_filter( $http, $filter ),
+         matrix_register_user_with_filter( $http, $filter ),
+      )->then( sub {
+         ( $user_a, $filter_id_a, $user_b, $filter_id_b ) = @_;
+
+         matrix_create_room( $user_a );
+      })->then( sub {
+         ( $room_id ) = @_;
+
+         matrix_invite_user_to_room( $user_a, $user_b, $room_id );
+      })->then( sub {
+         matrix_join_room( $user_b, $room_id );
+      })->then( sub {
+         matrix_sync( $user_b, $filter_id_b );
+      })->then( sub {
+         my ( $body ) = @_;
+
+         $next_b = $body->{next_batch};
+
+         matrix_send_room_text_message( $user_a, $room_id, body => "before" );
+      })->then( sub {
+         matrix_put_room_state( $user_a, $room_id,
+            type      => "a.madeup.test.state",
+            content   => { "my_key" => "before" },
+            state_key => "",
+         );
+      })->then( sub {
+         matrix_leave_room( $user_b, $room_id );
+      })->then( sub {
+         matrix_send_room_text_message( $user_a, $room_id, body => "after" );
+      })->then( sub {
+         matrix_put_room_state( $user_a, $room_id,
+            type      => "a.madeup.test.state",
+            content   => { "my_key" => "after" },
+            state_key => "",
+         );
+      })->then( sub {
+         matrix_sync( $user_b, filter => $filter_id_b );
+      })->then( sub {
+         my ( $body ) = @_;
```

Comment:
I'd rather avoid stuffing too much logging on failure. I found it very difficult to debug some of the tests due to overenthusiastic use of log_if_fail, since it made it more annoying to read the client logs. I'd suggest leaving it as is, and if someone finds it too difficult to debug without that logging then they can always add it themselves.